src/relax/op/tensor/transform.cc

```diff
@@ -0,0 +1,441 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file transform.cc
+ * \brief Transform operators.
+ */
+
+#include <numeric>
+
+#include "../op_common.h"
+
+namespace tvm {
+namespace relax {
+
+/* relax.reshape */
+TVM_REGISTER_NODE_TYPE(ReshapeAttrs);
+
+Expr MakeReshape(Expr data, ObjectRef new_shape) {
+  static const Op& op = Op::Get("relax.reshape");
+  if (const auto* expr = new_shape.as<ExprNode>()) {
+    return Call(op, {std::move(data), GetRef<Expr>(expr)}, Attrs(), {});
+  } else if (const auto* _array = new_shape.as<ArrayNode>()) {
+    ObjectPtr<ReshapeAttrs> attrs = make_object<ReshapeAttrs>();
+    attrs->new_shape = GetRef<Array<PrimExpr>>(_array);
+    return Call(op, {std::move(data)}, Attrs(attrs), {});
+  }
+  LOG(FATAL) << "Reshape only expects the input new shape to be either an Expr or an Array of "
+                "PrimExpr. However, the given one is "
+             << new_shape->GetTypeKey();
+  throw;
+}
+
+TVM_REGISTER_GLOBAL("relax.op.reshape").set_body_typed(MakeReshape);
+
+std::pair<PrimExpr, int> CheckNewShape(const Call& call, const BlockBuilder& ctx,
+                                       const Array<PrimExpr>& new_shape, bool allow_dim_inference) {
+  int dim_to_infer = -1;
+  int new_ndim = new_shape.size();
+  PrimExpr new_shape_prod = IntImm(DataType::Int(64), 1);
+  for (int i = 0; i < new_ndim; ++i) {
+    if (!new_shape[i]->dtype.is_int()) {
+      ctx->ReportFatal(Diagnostic::Error(call) << "Reshape requires the new shape values to be all "
+                                                  "integers. However, the give new shape is "
+                                               << new_shape);
+    }
+    const auto* int_len = new_shape[i].as<IntImmNode>();
+    if (int_len != nullptr && int_len->value == -1) {
+      if (!allow_dim_inference) {
+        ctx->ReportFatal(Diagnostic::Error(call)
+                         << "Reshape expects the new shape not to contain \"-1\" when the data "
+                            "shape is not known symbolically. However, the new shape is "
+                         << call->args[1]);
+      } else if (dim_to_infer != -1) {
+        ctx->ReportFatal(Diagnostic::Error(call)
+                         << "Reshape accepts at most one \"-1\" in the new shape. However, there "
+                            "are multiple \"-1\" in the given one "
+                         << call->args[1]);
+      }
+      dim_to_infer = i;
+    } else if (ctx->GetAnalyzer()->CanProveLess(new_shape[i], 1)) {
+      ctx->ReportFatal(Diagnostic::Error(call)
```

Comment:
This can be skipped, add a comment stating that we expect symbolic integers not to signal intent of -1.

Reply:
Rephrased the logic. Now we only check the “= -1 or ≥ 1” requirement for IntImms.